docs/fundamentals/code-analysis/quality-rules/ca2252.md

```diff
@@ -0,0 +1,64 @@
+helpviewer_keywords:
+- DetectPreviewFeaturesAnalyzer
+- CA2252
+author: pgovind
+ms.author: pgovind
+dev_langs:
+- CSharp
+- VB
+---
+# CA2252: Opt into Preview Features before using them
+
+| | Value |
+|-|-|
+| **Rule ID** |CA2252|
+| **Category** |[Microsoft.Usage](usage-warnings.md)|
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+A client uses preview APIs/types in their assembly without explicitly opting into it either locally or at the module/assembly level.
+
+## Rule description
+
+When an API/Assembly marked with the `RequiresPreviewFeaturesAttribute` attribute is consumed, this analyzer checks if the call site has opted into preview features. A call site has opted into preview features if it is within the scope of a `RequiresPreviewFeaturesAttribute` annotation or if it is part of an assembly/module that has already opted into preview features. For ex:
+
+> ![CA2252 - Detect Preview Feature Usage](media/ca2252-basic.png)
+
+Here, `Lib` is a preview type that is constructed in the `Main` method. `Main` itself is not annotated as a preview method, so diagnostics are produced on the 2 constructors calls inside `Main`.
+
+## How to fix violations
+
+There are 2 ways to fix violations:
+
+1. Bring a call site within the scope of an annotation by marking its parent with the `RequiresPreviewFeatures` annotation. In the example above, `APreviewMethod` is annotated with the `RequiresPreviewFeatures` attribute, so the analyzer ignores preview type usage inside `APreviewMethod`. It follows that callers of `APreviewMethod` will have to perform a similar exercise.
+
+
+2. If an assembly wishes, it can also opt into preview features at an assembly or module level. This indicates to the analyzer that preview type usage in the assembly is desired and, as a consequence, no errors will produced by this analyzer. This is the preferred way to consume preview dependencies. Setting the following property in a `.csproj` file will enable preview features inside the entire assembly:
+
+```csharp
+	<PropertyGroup>
+		<EnablePreviewFeatures>true</EnablePreviewFeatures>
+	</PropertyGroup>
+```
+
+
+## When to suppress errors
+
+This is only recommended for advanced use cases where diagnostics on APIs need to explicitly disabled and the user is willing to take on the responsiblity of marking preview APIs appropriately. Disable the analyzer locally as shown in the code snippet below. For ex: Consider a case where an existing type implements a new preview interface. Since the entire type cannot be marked as preview (for backwards compatibility), the diagnostic around the type definition can be disabled locally. Further, the preview interface implementations need to be marked as preview by the user. Now, the existing type can be used as before, but calls to the new interface methods will get diagnostics. System.Private.CoreLib.csproj uses this technique to expose generic math features on Int32, Double, Decimal etc.
```

Comment:
```suggestion
Suppressing errors from this rule is only recommended for advanced use cases where diagnostics on APIs need to explicitly disabled. In this case, you must be willing to take on the responsibility of marking preview APIs appropriately. For example, consider a case where an existing type implements a new preview interface. Since the entire type cannot be marked as preview (for backwards compatibility), the diagnostic around the type definition can be disabled locally. Further, you need to mark the preview interface implementations as preview. Now, the existing type can be used as before, but calls to the new interface methods will get diagnostics. *System.Private.CoreLib.csproj* uses this technique to expose generic math features on numeric types such as `Int32`, `Double`, and `Decimal`.

The following images show how to disable the CA2252 analyzer locally.  
```